### PR TITLE
TYP: get_indexer

### PR DESCRIPTION
**Pull request by jbrockmendel**
@simonjayhawkins im at a loss on why im getting a mypy complaint:

```
error: Overloaded function signatures 1 and 2 overlap with incompatible return types  [misc]
```

AFAICT they should be compatible.  Am I missing something obvious?

### CODE REVIEW

**Comment by jbrockmendel** [comment]
@simonjayhawkins gentle ping

**Comment by jreback** [comment]
can you rebase

**Comment by jbrockmendel** [comment]
cc @WillAyd any idea whats going on here?  my confusion only deepens when i try to do the same thing for Index.join

**Comment by jbrockmendel** [comment]
rebased, still no idea whats going on with mypy

**Comment by jbrockmendel** [comment]
@MarcoGorelli #40197 makes me think you have a better grasp of `@overload` than i do.  can you tell what im doing wrong here?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Hey @jbrockmendel ,

Yeah, I'm embarassed by how long I spent making sense of this in order to do #40200 and #40197 (maybe I should write a blog post* about it)

I explain it a bit more clearly in #https://github.com/pandas-dev/pandas/pull/40200#issue-584075687 , but in essence, you need an overload for each combination of keyword argument which _precedes_ the one you're overloading

Here, there are no keyword arguments preceding `unique`, so you'll just need overloads for:
- `Literal[True] = ...`
- `Literal[False]`
- `bool = ...`

What's confusing here is that the `bool` overload is needed - IMO, it shouldn't be, as the union of `Literal[True]` and `Literal[False]` should be regarded as the same as `bool`, but I opened an issue about that in mypy and have had no response: https://github.com/python/mypy/issues/10194

---

*just put something down while it's still fresh in my mind - and in case it's useful to you: https://medium.com/@m.e.gorelli/making-sense-of-typing-overload-437e6deecade

**Comment by MarcoGorelli** [comment]
Indeed, this seems to work:
```python
    @overload
    def _get_indexer_non_comparable(
        self, target: Index, method, unique: Literal[True] = ...
    ) -> np.ndarray:
        # returned ndarray is np.intp
        ...

    @overload
    def _get_indexer_non_comparable(
        self, target: Index, method, unique: Literal[False]
    ) -> tuple[np.ndarray, np.ndarray]:
        # both returned ndarrays are np.intp
        ...

    @overload
    def _get_indexer_non_comparable(
        self, target: Index, method, unique: bool = ...
    ) -> Union[tuple[np.ndarray, np.ndarray], np.ndarray]:
        ...
```

**Comment by jbrockmendel** [comment]
> Indeed, this seems to work:

It does, but AFAICT it isnt the 3rd `@overload` that is making a difference, its the fact that in the second one you wrote `unique: Literal[False]` instead of `unique: Literal[False] = ...`

**Comment by jbrockmendel** [comment]
> It does, but AFAICT it isnt the 3rd @overload that is making a difference, its the fact that in the second one you wrote unique: Literal[False] instead of unique: Literal[False] = ...

related: what if the argument for which we need to omit the `= ...` comes after another keyword-only argument?

**Comment by MarcoGorelli** [comment]
> > Indeed, this seems to work:
> 
> It does, but AFAICT it isnt the 3rd `@overload` that is making a difference, its the fact that in the second one you wrote `unique: Literal[False]` instead of `unique: Literal[False] = ...`

You need to omit `=...` for `Literal[False]` because the default is `Literal[True]`

> related: what if the argument for which we need to omit the = ... comes after another keyword-only argument?

See what I did for `set_axis` - you need to provide overloads (without `=...`) for each combination of keyword arguments which comes before the one you're overloading

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -5243,16 +5247,32 @@ def get_indexer_for(self, target, **kwargs) -> np.ndarray:
...
+        # returned ndarray is np.intp
+        ...
+
+    @overload
+    def _get_indexer_non_comparable(
+        self, target: Index, method, unique: Literal[False]
+    ) -> tuple[np.ndarray, np.ndarray]:
+        # both returned ndarrays are np.intp
+        ...
+
```
Comment (MarcoGorelli) [inline pandas/core/indexes/base.py:5271]: try doing `reveal_type` on a file with something like
```
unique: bool

reveal_type(_get_indexer_non_comparable(unique=unique))
```

I expect it tells you that no overload variant matches it - this is why we need an overload for the `unique: bool` case (unfortunately it's not currently inferred to be the union of the `Literal[True]` and `Literal[False]` return types)

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -5233,16 +5237,32 @@ def get_indexer_for(self, target, **kwargs) -> np.ndarray:
...
     @final
-    def _get_indexer_non_comparable(self, target: Index, method, unique: bool = True):
+    def _get_indexer_non_comparable(
+        self, target: Index, method, unique: bool = True
+    ) -> np.ndarray | tuple[np.ndarray, np.ndarray]:
```
Comment (MarcoGorelli) [inline pandas/core/indexes/base.py:5265]: This return type needs to be in an overload, not in the function signature itself - see this example:
```python
from typing import Literal, overload, Optional

@overload
def func(inplace: Literal[True]) -> None: ...

@overload
def func(inplace: Literal[False] = ...) -> int: ...

def func(inplace: bool = False) -> Optional[int]: ...

inplace: bool
reveal_type(func(inplace=inplace))
```
gives us
```python
main.py:12: error: No overload variant of "func" matches argument type "bool"
main.py:12: note: Possible overload variants:
main.py:12: note:     def func(inplace: Literal[True]) -> None
main.py:12: note:     def func(inplace: Literal[False] = ...) -> int
main.py:12: note: Revealed type is 'Any'
Found 1 error in 1 file (checked 1 source file)
```

However, if we make a separate overload:
```python
from typing import Literal, overload, Optional

@overload
def func(inplace: Literal[True]) -> None: ...

@overload
def func(inplace: Literal[False] = ...) -> int: ...

@overload
def func(inplace: bool = ...) -> Optional[int]: ...


def func(inplace: bool = False): ...

inplace: bool
reveal_type(func(inplace=inplace))
```
it (correctly) gives us
```python
main.py:16: note: Revealed type is 'Union[builtins.int, None]'
```

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py:5265]: sure, will update.  for my edification, why isn't this caught by the regular mypy run?

Comment (MarcoGorelli) [reply pandas/core/indexes/base.py:5265]: thanks - as far as I understand, if you have any overloads, then mypy will only look at them and not at the function signature

The reason you sometimes don't need to do this is that, according to [the docs](https://mypy.readthedocs.io/en/stable/more_types.html):

> if multiple variants match due to one or more of the arguments being a union, mypy will make the inferred type be the union of the matching variant returns:

So, if one overload is `str` and another is `bool` and you pass something that's `Union[str, bool]`, it'll infer the return type as being the union of the two overloads

Here, however, that is not the case, because it doesn't recognise `bool` as being the union of `Literal[True]` and `Literal[False]`

File: pandas/core/indexes/category.py [hunk 3]
```diff
@@ -494,22 +493,27 @@ def _maybe_cast_indexer(self, key) -> int:
...
+    def get_indexer_non_unique(self, target) -> tuple[np.ndarray, np.ndarray]:
+        # both returned ndarrays are np.intp
         target = ibase.ensure_index(target)
         return self._get_indexer_non_unique(target._values)
 
-    def _get_indexer_non_unique(self, values: ArrayLike):
+    def _get_indexer_non_unique(
+        self, values: ArrayLike
+    ) -> tuple[np.ndarray, np.ndarray]:
+        # both returned ndarrays are np.intp
```
Comment (MarcoGorelli) [inline pandas/core/indexes/category.py:516]: No need to add the return type to the function signature now, as the overloads will be used anyway, but I guess it's OK to keep

----

e.g.:
```python
from typing import overload, Literal, Optional, Union


@overload
def foo(bar: int) -> float: ...

@overload
def foo(bar: str) -> None: ...



def foo(bar: Union[str, int]): ...


bar: Union[int, str]
reveal_type(foo(bar))  # Revealed type is 'Union[builtins.float, None]'
```

Comment (simonjayhawkins) [reply pandas/core/indexes/category.py:516]: I think it is best to always include the types in the actual signature as well so that mypy checks the consistency of the function itself and reports an error with...

```
def foo(bar: Union[str, int]) -> float | None:
    return 42
```

Comment (MarcoGorelli) [reply pandas/core/indexes/category.py:516]: That's a good point, thanks!

**Comment by jbrockmendel** [comment]
Are any of the remaining comments actionable?

**Comment by jbrockmendel** [comment]
gentle ping @MarcoGorelli

**Review by MarcoGorelli — APPROVED** [review]
As far as I can tell, this is correct (though probably I don't know this part of the codebase well enough to merge)

----

The only part here I'm not totally comfortable commenting on is the change from `missing` to `ensure_platform_int(missing)`, as that involves making sense of `pxi.in` files, which I've not (yet) done. If that change is correct, then I'm pretty sure the rest is fine

File: pandas/core/indexes/base.py [hunk 4]
```diff
@@ -5217,7 +5221,7 @@ def get_indexer_non_unique(self, target):
         tgt_values = target._get_engine_target()
 
         indexer, missing = self._engine.get_indexer_non_unique(tgt_values)
-        return ensure_platform_int(indexer), missing
+        return ensure_platform_int(indexer), ensure_platform_int(missing)
```
Comment (MarcoGorelli) [inline pandas/core/indexes/base.py:5224]: perhaps you could just comment on why this change is necessary, and what it does?

**Comment by MarcoGorelli** [comment]
Thanks for explaining - thanks @jbrockmendel
